File: src/FeedbackBoard.Api/Services/FeedbackService.cs
using FeedbackBoard.Api.Hubs;
using FeedbackBoard.Api.Services.Interfaces;
using FeedbackBoard.Core.Entities;
using FeedbackBoard.Core.Events;
using FeedbackBoard.Core.Models;
using Mapster;
using Microsoft.AspNetCore.SignalR;

namespace FeedbackBoard.Api.Services;

public class FeedbackService(ICosmosDbService cosmosDb, IServiceBusPublisher serviceBus, IFeedbackMetadataService metadata,
    IHubContext<FeedbackHub> feedbackHub, ILogger<FeedbackService> logger) : IFeedbackService
{
    public async Task<FeedbackResponse> CreateFeedbackAsync(CreateFeedbackRequest request)
    {
        var feedback = request.Adapt<Feedback>();
        feedback.Id = Guid.NewGuid().ToString();
        feedback.Status = FeedbackStatusEnum.New;
        feedback.StatusHistory = new List<StatusChange>
        {
            new() { OldStatus = FeedbackStatusEnum.New, NewStatus = FeedbackStatusEnum.New, ChangedBy = request.UserId }
        };
        feedback.CreatedAt = DateTime.UtcNow;

        await cosmosDb.CreateFeedbackAsync(feedback);

        await serviceBus.PublishAsync(new FeedbackCreatedEvent
        {
            FeedbackId = feedback.Id,
            Title = feedback.Title,
            Description = feedback.Description,
            UserId = feedback.UserId,
            CreatedAt = feedback.CreatedAt
        });

        logger.LogInformation("Feedback created and event published: {Id}", feedback.Id);
        return await MapToResponseAsync(feedback);
    }

    public async Task<FeedbackResponse> ChangeStatusAsync(string feedbackId, FeedbackStatusEnum newStatus, string changedBy, string? reason = null)
    {
        var feedback = await cosmosDb.GetFeedbackAsync(feedbackId);
        if (feedback == null) throw new InvalidOperationException("Feedback not found");

        var oldStatus = feedback.Status;
        feedback.StatusHistory.Add(new StatusChange
        {
            OldStatus = oldStatus,
            NewStatus = newStatus,
            ChangedBy = changedBy,
            Reason = reason,
            Timestamp = DateTime.UtcNow
        });

        feedback.Status = newStatus;
        feedback.UpdatedAt = DateTime.UtcNow;
        if (newStatus == FeedbackStatusEnum.Completed) feedback.CompletedAt = DateTime.UtcNow;

        await cosmosDb.UpdateFeedbackAsync(feedback);

        await serviceBus.PublishAsync(new StatusChangedEvent
        {
            FeedbackId = feedback.Id,
            OldStatus = oldStatus,
            NewStatus = newStatus,
            ChangedBy = changedBy,
            Reason = reason
        });

        var response = await MapToResponseAsync(feedback);

        // signalR sending 
        await feedbackHub.Clients.Group("FeedbackWatchers").SendAsync("StatusChanged", new
        {
            feedbackId,
            status = response.StatusInfo
        });

        logger.LogInformation("Feedback status changed: {Id} {OldStatus} → {NewStatus}", feedback.Id, oldStatus, newStatus);
        return response;
    }

    public async Task<FeedbackResponse> VoteAsync(string feedbackId, string userId)
    {
        var feedback = await cosmosDb.GetFeedbackAsync(feedbackId);
        if (feedback == null) throw new InvalidOperationException("Feedback not found");

        if (!feedback.VoterIds.Contains(userId))
        {
            feedback.VoterIds.Add(userId);
            feedback.VoteCount = feedback.VoterIds.Count;
            await cosmosDb.UpdateFeedbackAsync(feedback);

            await serviceBus.PublishAsync(new FeedbackVotedEvent
            {
                FeedbackId = feedbackId,
                UserId = userId,
                NewVoteCount = feedback.VoteCount
            });
        }

        var response = MapToResponseAsync(feedback);

        await feedbackHub.Clients.Group("FeedbackWatchers").SendAsync("VoteUpdated", new
        {
            feedbackId,
            voteCount = feedback.VoteCount,
            hasVoted = true
        });

        return await MapToResponseAsync(feedback);
    }

    public async Task<FeedbackResponse?> GetFeedbackAsync(string id)
    {
        var feedback = await cosmosDb.GetFeedbackAsync(id);
        return feedback == null ? null : await MapToResponseAsync(feedback);
    }

    public async Task<List<FeedbackResponse>> GetFeedbacksByCategoryAsync(int categoryId)
    {
        var feedbacks = await cosmosDb.GetFeedbacksByCategoryAsync(categoryId);
        var responses = new List<FeedbackResponse>();
        foreach (var feedback in feedbacks)
        {
            responses.Add(await MapToResponseAsync(feedback));
        }
        return responses;
    }

    public async Task<List<FeedbackResponse>> GetFeedbacksByStatusAsync(FeedbackStatusEnum status)
    {
        var feedbacks = await cosmosDb.GetFeedbacksByStatusAsync(status);
        var responses = new List<FeedbackResponse>();
        foreach (var feedback in feedbacks)
        {
            responses.Add(await MapToResponseAsync(feedback));
        }
        return responses;
    }

    private async Task<FeedbackResponse> MapToResponseAsync(Feedback feedback)
    {
        var response = feedback.Adapt<FeedbackResponse>();

        response.StatusInfo = await metadata.GetStatusInfo(feedback);

        var category = await metadata.GetCategoryAsync(feedback.CategoryId);
        if (category != null)
        {
            response.CategoryName = category.DisplayName;
            response.CategoryIcon = category.Icon;
        }

        if (feedback.AiAnalysis?.SuggestedCategoryId != null)
        {
            var suggestedCategory = await metadata.GetCategoryAsync(feedback.AiAnalysis.SuggestedCategoryId.Value);
            response.SuggestedCategoryName = suggestedCategory?.DisplayName;
        }

        return response;
    }
}

File: frontend/src/api/feedbackService.ts
import axiosClient from "./axiosClient";

export const FeedbackStatus = {
  New: 1,
  UnderReview: 2,
  Planned: 3,
  InProgress: 4,
  Completed: 5,
  Declined: 6,
  Duplicate: 7,
} as const;

export type FeedbackStatus =
  (typeof FeedbackStatus)[keyof typeof FeedbackStatus];

export interface FeedbackStatusInfo {
  id: number;
  name: string;
  displayName: string;
  color: string;
}

export interface StatusChangeResponse {
  oldStatus: string;
  newStatus: string;
  changedBy: string;
  reason?: string;
  timestamp: string;
}

export interface FeedbackResponse {
  id: string;
  title: string;
  description: string;
  categoryId: number;
  categoryName?: string;
  categoryIcon?: string;
  statusInfo?: FeedbackStatusInfo;
  statusHistory: StatusChangeResponse[];
  sentiment?: string;
  suggestedCategoryName?: string;
  hasDuplicates: boolean;
  authorName: string;
  voteCount: number;
  hasVoted: boolean;
  createdAt: string;
  updatedAt?: string;
  completedAt?: string;
  attachmentUrls: string[];
}

export interface CreateFeedbackRequest {
  title: string;
  description: string;
  categoryId: number;
  userId: string;
  authorName: string;
}

export interface ChangeStatusRequest {
  newStatus: FeedbackStatus;
  changedBy: string;
  reason?: string;
}

export interface VoteRequest {
  userId: string;
}

export interface CategoryInfo {
  id: number;
  name: string;
  displayName: string;
  icon: string;
  description: string;
}

export const feedbackService = {
  createFeedback: async (
    data: CreateFeedbackRequest,
  ): Promise<FeedbackResponse> => {
    const response = await axiosClient.post<FeedbackResponse>(
      "/feedback",
      data,
    );
    return response.data;
  },

  getFeedbackById: async (id: string): Promise<FeedbackResponse> => {
    const response = await axiosClient.get<FeedbackResponse>(`/feedback/${id}`);
    return response.data;
  },

  getFeedbacksByCategory: async (
    categoryId: string,
  ): Promise<FeedbackResponse[]> => {
    const response = await axiosClient.get<FeedbackResponse[]>(
      `/feedback/category/${categoryId}`,
    );
    return response.data;
  },

  getFeedbacksByStatus: async (status: number): Promise<FeedbackResponse[]> => {
    const response = await axiosClient.get<FeedbackResponse[]>(
      `/feedback/status/${status}`,
    );
    return response.data;
  },

  getTopFeedbacks: async (count = 10): Promise<FeedbackResponse[]> => {
    const response = await axiosClient.get<FeedbackResponse[]>(
      `/feedback/top?count=${count}`,
    );
    return response.data;
  },

  changeFeedbackStatus: async (
    id: string,
    data: ChangeStatusRequest,
  ): Promise<FeedbackResponse> => {
    const response = await axiosClient.post<FeedbackResponse>(
      `/feedback/${id}/status`,
      data,
    );
    return response.data;
  },

  vote: async (id: string, data: VoteRequest): Promise<FeedbackResponse> => {
    const response = await axiosClient.post<FeedbackResponse>(
      `/feedback/${id}/vote`,
      data,
    );
    return response.data;
  },

  getStatuses: async (): Promise<FeedbackStatusInfo[]> => {
    const response =
      await axiosClient.get<FeedbackStatusInfo[]>(`/feedback/statuses`);
    return response.data;
  },

  getCategories: async (): Promise<CategoryInfo[]> => {
    const response =
      await axiosClient.get<CategoryInfo[]>(`/feedback/categories`);
    return response.data;
  },
};
